olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -155,18 +153,177 @@ client::CancellableFuture<DataResponse> VersionedLayerClientImpl::GetData(
+              CatalogVersionRequest().WithBillingTag(request.GetBillingTag()),
+              settings);
+
+          if (!response.IsSuccessful()) {
+            callback(response.GetError());
+            return EmptyResponse(PrefetchTileNoError());
+          }
+
+          request.WithVersion(response.GetResult().GetVersion());
+        }
+
+        const auto key = request.CreateKey(layer_id);
+        OLP_SDK_LOG_INFO_F(kLogTag, "PrefetchTiles: using key=%s", key.c_str());
+
+        // Calculate the minimal set of Tile keys and depth to
+        // cover tree.
+        auto sub_quads = repository::PrefetchTilesRepository::EffectiveTileKeys(
+            request.GetTileKeys(), request.GetMinLevel(),
+            request.GetMaxLevel());
+
+        if (sub_quads.empty()) {
+          OLP_SDK_LOG_WARNING_F(kLogTag,
+                                "PrefetchTiles: tile/level mismatch, key=%s",
+                                key.c_str());
+          callback({{ErrorCode::InvalidArgument, "TileKeys/levels mismatch"}});
+          return EmptyResponse(PrefetchTileNoError());
+        }
+
+        OLP_SDK_LOG_DEBUG_F(kLogTag, "PrefetchTiles, subquads=%zu, key=%s",
+                            sub_quads.size(), key.c_str());
+
+        // Now get metadata for subtiles using QueryApi::QuadTreeIndex
+        auto sub_tiles =
+            prefetch_repo->GetSubTiles(layer_id, request, sub_quads, context);
+
+        if (!sub_tiles.IsSuccessful()) {
+          callback(sub_tiles.GetError());
+          return EmptyResponse(PrefetchTileNoError());
+        }
+
+        const auto& tiles_result = sub_tiles.GetResult();
+        if (tiles_result.empty()) {
+          OLP_SDK_LOG_WARNING_F(
+              kLogTag, "PrefetchTiles: subtiles empty, key=%s", key.c_str());
+          callback({{ErrorCode::InvalidArgument, "Subquads retrieval failed"}});
+          return EmptyResponse(PrefetchTileNoError());
+        }
+
+        OLP_SDK_LOG_INFO_F(kLogTag, "Prefetch start, key=%s, tiles=%zu",
+                           key.c_str(), tiles_result.size());
+
+        // Once we have the data create for each subtile a task and push it
+        // onto the TaskScheduler. One additional last task is added which
+        // waits for all previous tasks to finish so that it may call the user
+        // with the result.
+        auto futures = std::make_shared<std::vector<PrefetchResultFuture>>();
+        std::vector<CancellationContext> contexts;
+        contexts.reserve(tiles_result.size() + 1u);
+        auto it = tiles_result.begin();
+
+        while (!context.IsCancelled() && it != tiles_result.end()) {
+          auto promise = std::make_shared<PrefetchResultPromise>();
+          auto flag = std::make_shared<std::atomic_bool>(false);
+          futures->emplace_back(std::move(promise->get_future()));
+
+          auto tile = it->first;
+          auto handle = it->second;
+          auto context_it = contexts.emplace(contexts.end());
```

Comment:
Yes

Reply:
I didn't figured out that when was reading this first time - I suggest to pass `CancellationToken` explicitly, as it is more obvious for the readers / maintainers. 

Also, we can use just `emplace_back`: `contexts.emplace_back(CancellationToken{});`
and instead iterator we can use: `contexts.back()` during call to AddTask;

Reply:
Theoretically if you use `emplace_back()` and the constructor is empty then you should do
```cpp
contexts.emplace_back();
```

else you are using copy/move constructor from an temporary var. So this is why i didn't use emplace_back() as it would have been empty, and the code would look like this:

```cpp
contexts.emplace_back();
context = contexts.back();
```
Does this look more clear? If yes, then this is exactly what `auto context_it = contexts.emplace(contexts.end());` does, only in one line of code.